modules/openxr/extensions/openxr_vulkan_extension.cpp

```diff
@@ -37,21 +37,24 @@
 #include "servers/rendering/rendering_server_globals.h"
 #include "servers/rendering_server.h"
 
-OpenXRVulkanExtension::OpenXRVulkanExtension(OpenXRAPI *p_openxr_api) :
-		OpenXRGraphicsExtensionWrapper(p_openxr_api) {
+OpenXRVulkanExtension::OpenXRVulkanExtension() {
 	VulkanContext::set_vulkan_hooks(this);
-
-	request_extensions[XR_KHR_VULKAN_ENABLE2_EXTENSION_NAME] = nullptr; // must be available
-
-	ERR_FAIL_NULL(openxr_api);
 }
 
 OpenXRVulkanExtension::~OpenXRVulkanExtension() {
 	VulkanContext::set_vulkan_hooks(nullptr);
 }
 
+HashMap<String, bool *> OpenXRVulkanExtension::get_request_extensions() {
+	HashMap<String, bool *> request_extensions;
+
+	request_extensions[XR_KHR_VULKAN_ENABLE2_EXTENSION_NAME] = nullptr; // must be available
+
+	return request_extensions;
+}
+
 void OpenXRVulkanExtension::on_instance_created(const XrInstance p_instance) {
-	ERR_FAIL_NULL(openxr_api);
+	ERR_FAIL_NULL(OpenXRAPI::get_singleton());
```

Comment:
In this case the checks are important though admittedly we're nitpicking here. OpenXRAPI will only exist when OpenXR is enabled and when we're in runtime. In the editor OpenXRAPI (unless the VR editor is enabled) won't be instantiated and can thus be NULL. But you are right, so far it's a bit hit and miss where we do and don't check it, and the chance of someone actually implementing things wrong are near zero because we don't instantiate the UI parts if OpenXR is turned off.

Still it's better to be safe and sorry with public APIS.

As for storing the pointer in a variable, I always use to do that but I'm on the fence on that one atm. This did remind me I needed to change the singleton call to be inlined, but even if not modern compilers will pick up on the function just returning a static value and optimizing it to just directly refer to that static variable.

I do admit things get a little more verbose this way...

Reply:
Imho defensive programming should always remain stable, even if things outside of the method change. `get_singleton` might theoretically change, so I'd argue it is better to not rely on compiler optimizations caused by its inline implementation.

But I do agree we're nitpicking, and it is not terribly important either way. It just felt to me as if you search/replaced `openxr_api` with `OpenXRAPI::get_singleton()` and the checks were unintentionally left over.

Reply:
Totally agree with you there, but I think that ship sailed a long time ago. Godot has many places where singletons are used in this manner.

And like I said, I'm on the fence about some of this. There really is no performance to gain here so code readability is an important factor too, and there really isn't much reason to think this code would ever be unintentionally called. I might still change it :)